### PR TITLE
Extension to CMake overhaul to completely avoid Makefile.kokkos

### PR DESCRIPTION
**Pull request by jjwilke**

### CODE REVIEW

**Comment by crtrott** [comment]
Does this supersede #2104 ?

**Comment by jjwilke** [comment]
Probably? It depends on how incrementally you want to roll out the CMake.

**Comment by dalg24** [comment]
> Does this supersede #2104 ?

No it would make no sense and we would lose the feedback that was given on the other PR.

That said I have no objection reviewing the two commits introduced by this PR separately.  It might get complicated to keep both in sync though.

**Comment by jjwilke** [comment]
> > Does this supersede #2104 ?
> 
> No it would make no sense and we would lose the feedback that was given on the other PR.
> 
> That said I have no objection reviewing the two commits introduced by this PR separately. It might get complicated to keep both in sync though.

I would recommend someone reviews this pull request and "approves" the Makefile.kokkos logic. I'll then push these changes back into the cmake-overhaul branch so the new commits get appended to the previous pull request and we don't lose the discussion.

File: cmake/kokkos_arch.cmake [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+IF (KOKKOS_ARCH_PASCAL60)
+  GLOBAL_APPEND(KOKKOS_CUDA_OPTIONS "${CUDA_ARCH_FLAG}=sm_60")
+ENDIF()
+
+IF (KOKKOS_ARCH_PASCAL61)
+  GLOBAL_APPEND(KOKKOS_CUDA_OPTIONS "${CUDA_ARCH_FLAG}=sm_61")
+ENDIF()
+
+IF (KOKKOS_ARCH_VOLTA70)
+  GLOBLA_APPEND(KOKKOS_CUDA_OPTIONS "${CUDA_ARCH_FLAG}=sm_70")
```
Comment (masterleinad) [inline cmake/kokkos_arch.cmake]:
```suggestion
  GLOBAL_APPEND(KOKKOS_CUDA_OPTIONS "${CUDA_ARCH_FLAG}=sm_70")
```

**Comment by masterleinad** [comment]
Running
```
cmake -DCMAKE_CXX_COMPILER=nvcc_wrapper ..
```
gives me:
```
-- The CXX compiler identification is GNU 7.4.0
-- Check for working CXX compiler: nvcc_wrapper
-- Check for working CXX compiler: nvcc_wrapper -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - failed
-- The project name is: Kokkos
-- KOKKOS_ENABLE_SERIAL=ON
-- KOKKOS_ENABLE_PROFILING=ON
CMake Error at cmake/kokkos_cxx.cmake:199 (kokkos_set_gnu_flags):
  kokkos_set_gnu_flags Function invoked with incorrect arguments for function
  named: kokkos_set_gnu_flags
Call Stack (most recent call first):
  cmake/kokkos_tribits.cmake:188 (INCLUDE)
  CMakeLists.txt:38 (KOKKOS_SETUP_BUILD_ENVIRONMENT)


CMake Error at cmake/kokkos_cxx.cmake:206 (CHECK_CXX_COMPILER_FLAG):
  CHECK_CXX_COMPILER_FLAG Macro invoked with incorrect arguments for macro
  named: CHECK_CXX_COMPILER_FLAG
Call Stack (most recent call first):
  cmake/kokkos_tribits.cmake:188 (INCLUDE)
  CMakeLists.txt:38 (KOKKOS_SETUP_BUILD_ENVIRONMENT)


CMake Error at cmake/kokkos_cxx.cmake:208 (MESSAGE):
  GNU did not accept .  You likely need to reduce the level of the C++
  standard from
Call Stack (most recent call first):
  cmake/kokkos_tribits.cmake:188 (INCLUDE)
  CMakeLists.txt:38 (KOKKOS_SETUP_BUILD_ENVIRONMENT)


-- Configuring incomplete, errors occurred!
See also "/tmp/ArborX/my_docker/kokkos/build_gcc/CMakeFiles/CMakeOutput.log".
See also "/tmp/ArborX/my_docker/kokkos/build_gcc/CMakeFiles/CMakeError.log".
```

File: cmake/kokkos_cxx.cmake [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+#Intel: Intel compiler versions 12.1 through 17.0.
+#
+#Cray: Cray Compiler Environment version 8.1 through 8.5.8.
+#PGI: PGI version 12.10 through 17.5.
+#XL: IBM XL version 10.1 through 13.1.5.
+
+kokkos_option(CXX_STANDARD "" STRING "The C++ standard for Kokkos to use: c++11, c++14, or c++17")
+
+IF (NOT KOKKOS_CXX_STANDARD AND NOT CMAKE_CXX_STANDARD)
+  SET(KOKKOS_CXX_STANDARD "11" CACHE STRING "C++ standard")
```
Comment (masterleinad) [inline cmake/kokkos_cxx.cmake]:
```suggestion
  SET(KOKKOS_CXX_STANDARD "11" CACHE STRING "C++ standard" FORCE)
```

File: cmake/kokkos_cxx.cmake [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+ELSEIF(NOT KOKKOS_CXX_STANDARD)
+  SET(KOKKOS_CXX_STANDARD ${CMAKE_CXX_STANDARD} CACHE STRING "C++ standard")
```
Comment (masterleinad) [inline cmake/kokkos_cxx.cmake]:
```suggestion
  SET(KOKKOS_CXX_STANDARD ${CMAKE_CXX_STANDARD} CACHE STRING "C++ standard" FORCE)
```

File: cmake/kokkos_arch.cmake [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+  )
+ENDIF()
+
+
+IF (KOKKOS_ARCH_KEPLER30)
+  GLOBAL_APPEND(KOKKOS_CUDA_OPTIONS "${CUDA_ARCH_FLAGS}=sm_30")
+ENDIF()
+
+IF (KOKKOS_ARCH_KEPLER32)
+  GLOBA_APPEND(KOKKOS_CUDA_OPTIONS "${CUDA_ARCH_FLAG}=sm_32")
```
Comment (masterleinad) [inline cmake/kokkos_arch.cmake]:
```suggestion
  GLOBAL_APPEND(KOKKOS_CUDA_OPTIONS "${CUDA_ARCH_FLAG}=sm_32")
```

File: cmake/kokkos_functions.cmake [hunk 5]
```diff
@@ -2,344 +2,42 @@
...
+    #CMAKE Makes this impossible to distinguish something already in cache
+    #AND SOMthing given explicitly on the command line
+    #AT THIS point, we have no choice but to accept the Camel value and print a warning
+    IF (NOT ${CAMEL_NAME} STREQUAL ${UC_NAME})
+      MESSAGE(WARNING "Overriding ${UC_NAME}=${${UC_NAME}} with ${CAMEL_NAME}=${${CAMEL_NAME}}")
+    ENDIF()
+    #I HAVE to accept the Camel case value - really no choice here - force it
+    SET(${UC_NAME} ${${CAMEL_NAME}} CACHE ${TYPE} ${DOCSTRING} FORCE)
+  ELSE() #GReat, no camel case names - nice and simple
+    SET(${UC_NAME} ${DEFAULT} CACHE ${TYPE} ${DOCSTRING})
```
Comment (masterleinad) [inline cmake/kokkos_functions.cmake]:
```suggestion
    SET(${UC_NAME} ${DEFAULT} CACHE INTERNAL ${DOCSTRING})
```

I guess you don't want this to show up in `ccmake`.

**Comment by masterleinad** [comment]
It seems that specifying `Kokkos_DIR` is still not sufficient for `CMake` to find it.
`KokkosConfig.cmake` exists in `${INSTALL}/lib/cmake/` and in `${INSTALL}/lib/CMake/Kokkos` but not in `${INSTALL}/lib/cmake/Kokkos` where `CMake` would find it.

**Comment by masterleinad** [comment]
Then, I also see
```
CMake Error at kokkos-install-nvcc/lib/cmake/Kokkos/KokkosTargets.cmake:102 (message):
  The imported target "Kokkos::kokkoscore" references the file
     "/tmp/ArborX/kokkos-install-nvcc/lib/lib/libkokkoscore.a"
  but this file does not exist.  Possible reasons include:
  * The file was deleted, renamed, or moved to another location.
  * An install or uninstall procedure did not complete successfully.
  * The installation package was faulty and contained
     "/tmp/ArborX/kokkos-install-nvcc/lib/cmake/Kokkos/KokkosTargets.cmake"
  but not all the files it references.
```
after copying `${INSTALL}/lib/cmake/` to `${INSTALL}/lib/cmake/Kokkos`.

**Comment by jjwilke** [comment]
> It seems that specifying `Kokkos_DIR` is still not sufficient for `CMake` to find it.
> `KokkosConfig.cmake` exists in `${INSTALL}/lib/cmake/` and in `${INSTALL}/lib/CMake/Kokkos` but not in `${INSTALL}/lib/cmake/Kokkos` where `CMake` would find it.

I'm not sure what you mean by CMake finding things. `Kokkos_DIR` has to be the folder containing `KokkosConfig.cmake`

**Comment by masterleinad** [comment]
> I'm not sure what you mean by CMake finding things. `Kokkos_DIR` has to be the folder containing `KokkosConfig.cmake`

Basically, https://dalg24.github.io/kokkos/#/6/4 and https://dalg24.github.io/kokkos/#/6/5.
I would naturally assume that `Kokkos_DIR` is the folder `Kokkos` is installed to. This works for other packages really well. Installing `KokkosConfig.cmake` into `${INSTALL}/lib/Kokkos/CMake` would work as well for finding `Kokkos` using `Kokkos_DIR=$INSTALL` as an environment variable (or appending `$INSTALL` to `CMAKE_PREFIX_PATH`).

**Comment by jjwilke** [comment]
@masterleinad One would naturally assume. And this is the behavior that would make sense to me. Unfortunately `<PackageName>_DIR` only causes CMake to look in that folder.

Possibly you know more than I do and there is a policy you can set to change this behavior?

I know you can use:
````
cmake_policy(SET CMP0074 NEW)
find_package(...)
````
and then use `-DKokkos_ROOT=<InstallPrefix>`.  This will then find the package file in the `lib/cmake/Kokkos` folder.  So maybe this is what you were referring to - so point taken. I'm developing on Mac which means this accidentally works because Darwin in its infinite wisdom ignores case.  Anyhow, issue #2162 opened to address this.

File: cmake/kokkos_functions.cmake (hunk 5 above)
Comment (jjwilke) [reply cmake/kokkos_functions.cmake]: agreed

File: cmake/kokkos_arch.cmake (hunk 4 above)
Comment (jjwilke) [reply cmake/kokkos_arch.cmake]: oops

File: cmake/kokkos_cxx.cmake (hunk 3 above)
Comment (jjwilke) [reply cmake/kokkos_cxx.cmake]: yes, you are correct

File: cmake/kokkos_cxx.cmake (hunk 2 above)
Comment (jjwilke) [reply cmake/kokkos_cxx.cmake]: yep, will fix, thank you

**Comment by crtrott** [comment]
I agree with 3.0 has a CMake system which doesn't call the pure Makefiles at all. We just need to get a ton of testing done to make sure we don't overlook too much. I am not super super worried if we do initially though. In Trilinos we can hot fix it, and outside people can go back to 2.9 and have all the same code features available if necessary.

**Comment by jjwilke** [comment]
Closing the discussion here. 3.0 release will be based on cmake-overhaul branch. Changes from this branch have been merged. Refer to #2104.

**Pull request closed without merging**